### ISSUE TITLE
P3 IDE on Windows doesn't support screen readers

### ISSUE THREAD

**Issue by toolness** [processing/processing#4312]
Hello!

I am working with @cleezyitp and the Processing Foundation to attempt to make the Processing 3 IDE work with--at the very least--screen-readers on Windows.

Currently, the P3 IDE doesn't work well with screen-readers on Windows: on both [NVDA](http://www.nvaccess.org/) and [JAWS](http://www.freedomscientific.com/Products/Blindness/JAWS), the screen reader is only able to see the title bar of the P3 IDE window.

As @Manindra29 told us, a number of other widgets, such as the toolbar and text-editing widget, are actually custom components that may or may not already support the [`Accessible`](https://docs.oracle.com/javase/7/docs/api/javax/accessibility/Accessible.html) interface; it's understandable that screen-readers can't see these.

However, other parts of the P3 IDE that *should* be accessible are still invisible to screen readers.  For example, the menu bar, which I believe is a standard Java Swing [`JMenuBar`](https://docs.oracle.com/javase/7/docs/api/javax/swing/JMenuBar.html) widget which implements the `Accessible` interface, can't be seen by JAWS or NVDA.

Upon further research, I learned about the [Java Access Bridge](http://www.oracle.com/technetwork/articles/javase/index-jsp-136191.html), which seemed like it might help:

> Java Access Bridge is a technology that exposes the Java Accessibility API in a Microsoft Windows DLL, enabling Java applications and applets that implement the Java Accessibility API to be visible to assistive technologies on Microsoft Windows systems.

I noticed that the P3 IDE on Windows uses something called [launch4j](http://launch4j.sourceforge.net/docs.html) to bundle the Java runtime along with the IDE so end-users don't need to download and install the JRE themselves. It seems the Java Access Bridge is actually present in this bundled JRE, in `java\bin\jabswitch.exe`, but [enabling and running this Java Access Bridge](http://docs.oracle.com/javase/7/docs/technotes/guides/access/enable_and_test.html) doesn't seem to help. It appears this problem with running `jabswitch.exe` in bundled JREs has been [documented on StackOverflow](http://stackoverflow.com/questions/20655101/jabswitch-not-working-in-bundled-jre), but there don't seem to be any good solutions to it yet.

That's as far as we've gotten on solving this mystery so far; we'll update this issue as we learn more.

(I hope it's OK that we're using a GitHub Issue to document this--if not, just let us know and we'll move it to somewhere else more appropriate!)

**Comment by toolness** [comment]
**Update:** The StackOverflow thread [*does* mention one workaround](http://stackoverflow.com/questions/20655101/jabswitch-not-working-in-bundled-jre#comment48963857_20655101) whereby downloading and installing the JRE system-wide and running the Java Access Bridge on it allows screen-readers to penetrate apps that use private bundled JREs (e.g., the P3 IDE). It's not a particularly user-friendly solution, but at least it might give blind users a way to use Processing!

**Comment by toolness** [comment]
Just verified that globally installing the Java SE Runtime Environment (Version 8 Build 73 / build 1.8.0_73-b02, in my case) and following the steps outlined in [Enabling and Testing Java Access Bridge](http://docs.oracle.com/javase/7/docs/technotes/guides/access/enable_and_test.html) makes Processing's standard Swing components visible to NVDA! Specifically, I followed these steps to enable the JAB:

1. Go to **Start > Control Panel > Ease of Access > Ease of Access Center**.
2. Select **Use the computer without a display**.
3. In the section **Other programs installed**, select the check box **Enable Java Access Bridge** (you may have to scroll down).

So this makes a bunch of things visible to the screen reader: the menu bar, a number of standard dialogs, and the console area, to name a few. However, as @Manindra29 predicted, the toolbar (with its "stop" and "run") buttons is invisible to the screen reader, and while the code editing area *does* announce keypresses while it's focused, there's no announcement of cursor or selection changes (as there is with the standard Notepad app that comes with Windows, and probably with other standard text areas).

**Comment by Manindra29** [comment]
Processing uses a bundled JRE to avoid the hassle for users to install a JRE themselves and any incompatibility issues that may arise with that. Since this is a very specific scenario, it may be worth considering. @benfry may want to chip in here.

As for the toolbar, I believe it shouldn't be a too much work to reimplement it with a swing components version with the same functionality and look and feel. The images on the buttons are custom designed. As for the text area, it should be possible to add the unimplemented methods from the [Accessible Component](https://docs.oracle.com/javase/7/docs/api/javax/accessibility/AccessibleComponent.html) interface to support the full range of text area accessibility features.

**Comment by benfry** [comment]
Probably better to implement the correct interfaces in the custom components rather than re-implement the components using Swing. Swing is kind of a nightmare, which is why `JButton` et al weren't overridden in the first place. (Witness the difficulty of getting the Contribution Manager to not look like a complete mess over the last year or so.) 

On the other hand, it might be better to create a simpler/accessible IDE that's just a front-end to the command line version. A lot of what's in the IDE is really predicated on sighted users, so you might consider whether retrofitting is the right direction. I can't say offhand, but I'd give some thought to what you need from an IDE for an audience that's using these accessibility features, and work from there. 

Happy to weigh in where helpful. Seems like you folks should set up a repo or a fork that you can work from and file issues there, since this isn't so much an 'issue' for the main project.

**Comment by CleezyITP** [comment]
@Manindra29, @toolness and @benfry , I have just had a conversation with the incredibly knowledgable @sinabahram, with respect to the processing  IDE’s accessibility! 

All things considered, we agree with Ben that making a standalone editor might be the easiest and most impactful way to bring accessibility to Processing. Swing has done all the heavy lifting for us and if we could put together something that is backed to processing like a smart editor app we can use SWT and base it on Eclipse. Can we talk more about this  @Manindra and @toolness?

**Comment by sinabahram** [comment]
Nice to meet everyone, albeit virtually. I’d also like to mention one tool that may be helpful in investigating the accessibility information exposed via the Java Access Bridge on Windows:

https://github.com/google/access-bridge-explorer
